### PR TITLE
feat(sync) Retry sync operations

### PR DESCRIPTION
**Pull request by cwurm**
Admin API calls to Kong can fail for a number of reasons, some permanent (e.g. configuration syntax error). Others are transient, e.g.:
1. Kong - or rather the underlying configuration database - gets overloaded with too many requests and will start returning `500` errors. I found this almost certain to happen with very large configurations of thousands or tens of thousands of entities.
2. Plugins can have race conditions, e.g. on setup when several plugin instances are created at once by decK. I have seen this just today.

When these transient errors happen a decK run will abort since the current behavior is for each decK worker to stop when it encounters any error whatsoever. Having to restart decK for transient reasons either manually or automatically is not ideal - esp. when it's a large configuration it will have to re-query all configuration, and potentially re-sync a lot of content (e.g. when using a decK config without default values specified).

This PR adds a `--retry` and `--retry-delay` parameter to the `deck sync` command. Default values are `1` retry and no delay. This should solve the most obvious transient problems by default without adding a lot of execution time (adding a delay can, since then each worker might have to wait for that time before shutting down, which could be a long time if there are a lot of errors).

To test this, what I've seen work well is to run deck with a very high number of workers, e.g. `deck sync -s kong.yaml --parallelism 10000 --retries 0`. With any reasonably large Kong configuration, this will likely fail (it does for me with about 120 entities). With `deck sync -s kong.yaml --parallelism 10000 --retries 1 --retry-delay 2` (retry delay set to 2s here) it should work.

### CODE REVIEW

File: cmd/sync.go
Comment (hbagdi) [inline cmd/sync.go:64]: Please do not make retry delay configurable (yet).
Why?
1. Keeping configuration parameters to minimum is important for a good user experience.
2. Configuring delay in such cases seldom adds value. The best practice in this area is to perform an exponential backoff delay which saves the day in a lot of such failure scenarios. Thundering herd is a real problem especially when default parallelism is 10.

Hardcode a delay scheme using a good exponential backoff algorithm library.

File: cmd/sync.go
Comment (hbagdi) [inline cmd/sync.go:62]: I'd advise even against making retries configurable in the beginning. I do see value here but I'd like to see users asking for it before we add it. Having an internally hardcoded value of `5` retries should do it for most users.

**Review by hbagdi — CHANGES_REQUESTED** [review]
Apologies for the delay in review here.

**Comment by cwurm** [comment]
@hbagdi Thanks for the review, I finally had time to look at this again and rewrote the implementation to use the `cenkalti/backoff` exponential backoff library.

The tricky part is that to have a successful default retry policy the potential maximum backoff is very long. In my tests, to be able to load several thousand Kong entities with `--parallelism 100` (trying to approximate a very large, complex and frequently automatically changed Kong deployment) a backoff of up to 10 seconds was required.

This is annoying when the errors are not retryable, e.g. wrong plugin configuration, access not authorized, etc. If all classes of errors are retried, decK will run through all retries even when the error is permanent and take a long time (15-20 seconds) doing nothing before finally exiting with the error.

So I needed to add some logic for which errors are retryable. In my tests, Kong will usually return a `500 Internal Server Error` in those cases, e.g. when its database connection gets overloaded. Bad plugin configuration is a `400 Bad Request`, missing access rights is `401 Unauthorized`, etc.

Unfortunately, `go-kong` does not currently expose the status code in an easily consumable way. We could change it to do that, what do you think? For now I made it work by checking the error message returned from go-kong.

File: cmd/sync.go
Comment (cwurm) [reply cmd/sync.go]: Makes sense, I changed to an exponential backoff.

File: cmd/sync.go
Comment (cwurm) [reply cmd/sync.go]: Makes sense, I've experimented with some values for the updated implementation with exponential backoff and came up with 3 retries after around 1 second, 3 seconds, and 9 seconds (randomized to avoid the thundering herd). Let me know what you think.

File: diff/diff.go
Comment (cwurm) [reply diff/diff.go:374]: Let me know what you think of the updated implementation. Still in `eventLoop()`, I guess could move it to `handleEvent()` though I think they'd look pretty much the same.

File: solver/solver.go
Comment (hbagdi) [inline solver/solver.go:68]: Can we make this a little more robust by checking against a regex instead to account for 501/502/503 or rather 5xx?

I've also opened up https://github.com/Kong/go-kong/issues/10 to make this better in future.

File: diff/diff.go
Comment (hbagdi) [reply diff/diff.go:374]: I think it is best to move it to handleEvent as it keeps the code testable and keeps the responsibilities of functions separate.

File: diff/diff.go
Comment (hbagdi) [inline diff/diff.go:383]: Please declare backoffObject as a global variable and initialize it in an `init()`. That keeps this function (or rather handleEvent fn after refactoring) cleaner. Readability and maintainability in this part of the code is more important than usual because it is easy to trip over.

File: diff/diff.go
Comment (hbagdi) [inline diff/diff.go:386]: Can we please emit a log line when an event is being retried? That would help quite a bit with debugging.

**Review by hbagdi — CHANGES_REQUESTED** [review]

File: solver/solver.go
Comment (cwurm) [reply solver/solver.go:68]: I'm not sure if anything other than 500 can be a transient error? The trouble with retrying a permanent error would be that the decK run will take a long time while it is retrying every request the maximum number of times. So far I've only seen 500 error be transient, e.g. when the database was overloaded with too many requests. So would it not make sense to be conservative here?

File: diff/diff.go
Comment (cwurm) [reply diff/diff.go:386]: I agree. However, I don't see any debug logging facility in decK? We have the print package with things like CreatePrintln/DeletePrintln, but those will always print. I imagine that ideally this kind of debug information could be turned on or off, e.g. with a verbose or debug flag.

File: solver/solver.go
Comment (hbagdi) [reply solver/solver.go:68]: In practice, a lot of times there is another hop or proxy between decK and the Admin endpoint of Kong. That intermediate layer can send a 502 or 503 intermittently. This is certainly a hypothetical scenario and we can defer the change.

File: diff/diff.go
Comment (hbagdi) [reply diff/diff.go:386]: There is a verbose flag in decK. You are right that the machinery around debug logging is lacking and there is not much that can be done easily. Tracking in #248 and #249.

File: diff/diff.go
Comment (cwurm) [reply diff/diff.go]: done

File: solver/solver.go
Comment (cwurm) [reply solver/solver.go:68]: Yeah, I see how there can be other transient errors. I like the idea of keeping the change small and restricted to just the specific case where we know decK does not do what a user would expect (load a large config with 1000s/10000s of Kong entities). We can always easily expand the reach if it comes up?

File: diff/diff.go
Comment (cwurm) [reply diff/diff.go:386]: Makes sense, I moved it.

**Review by hbagdi — APPROVED** [review]

**Comment by hbagdi** [comment]
Holding this merge until #250 gets merged in and released. That keeps the release process simpler.

**Comment by hbagdi** [comment]
@mflendrich Can I request you review this patch?

File: solver/solver.go
Comment (mflendrich) [inline solver/solver.go:68]: - Instead of parsing the error message, please update go-kong to `v0.14.0` and extract `Code()` from the [`APIError` type](https://github.com/Kong/go-kong/blob/f71c83b51f21bf129b63a60aa88447203d6bfb8b/kong/error.go#L8).
- Then, instead of literal `500`, use `http.StatusInternalServerError`

Comment (cwurm) [reply solver/solver.go]: Glad to see the addition to go-kong, done.

File: diff/diff.go
Comment (cwurm) [reply diff/diff.go:29]: So this is just implementing the review from @hbagdi here: https://github.com/Kong/deck/pull/226#discussion_r535421432

Happy to do this any way that works, just let me know.

Comment (mflendrich) [reply diff/diff.go:29]: Oh, apologies about contradicting review comments :joy:. I'll get this sorted out.

Comment (mflendrich) [reply diff/diff.go:29]: Sorry about changing expectations on our side. I have sorted this out with @hbagdi and we'd like to go for a helper function.

I meant to save you a round-trip by pushing the change to your PR branch but it seems that GH doesn't allow me to push. Therefore, please cherry-pick 35bccc0a3bfe5d3583888db75133ea88f40c8a74 or make an equivalent change yourself.

Thanks!

File: diff/diff.go
Comment (mflendrich) [inline diff/diff.go:408]:
```suggestion

```

**Review by mflendrich — COMMENTED** [review]
Final nit (+ the backoff helper function comment), otherwise LGTM. Thanks for your contribution!

**Comment by cwurm** [comment]
Thanks @mflendrich, helper function implemented!

**Review by mflendrich — APPROVED** [review]

**Comment by cal5barton** [comment]
Is there any reason why the flags are not listed in the general documentation?

**Comment by rainest** [comment]
As best I can tell, per https://github.com/Kong/deck/pull/226#discussion_r513654575 we actually removed the flags from the final version, but forgot to omit them from the commit message. The actual functionality just always retries with a set count and delay configuration.

Asked around to see if someone else with more history on this feature knows more, but that's what looks to be the case based on what I can glean from the PR discussion.

**Comment by cal5barton** [comment]
Right now I'm experiencing 502 errors ( #526 ) and was hoping that this retry functionality would solve something like that.

**Pull request merged**
